### PR TITLE
feat: Add system tool calls to the chat endpoint

### LINKED ISSUE #1285: [Feature]: (API Calls | Integrations | System Calls) Need to be Supported in Sessions Chat
### 🔖 Feature description

See #677 

### 🎤 Why is this feature needed ?

_No response_

### ✌️ How do you aim to achieve this?

_No response_

### 👀 Have you searched issues and PRs to see if this feature request has been raised before?

- [x] I checked and didn't find similar issue

### PR DESCRIPTION
**Pull request by whiterabbit1983**
### **User description**

Closes #1285 

---
# EntelligenceAI PR Summary 
 The recent update introduces a new utility function, `eval_tool_calls`, in `tools.py` to enhance the evaluation of tool calls within chat sessions. The `chat` function in `chat.py` has been updated to utilize this utility, replacing direct calls to `litellm.acompletion`. This change aims to improve code modularity and streamline the integration of tool responses in chat sessions.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced `tool_calls_evaluator` to streamline tool call handling in chat sessions.

- Added a new utility function `call_tool` for executing system tool calls.

- Implemented extensive unit tests for `tool_calls_evaluator` and `call_tool`.

- Enhanced modularity and maintainability of the chat endpoint.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>chat.py</strong><dd><code>Refactor chat endpoint to use tool evaluator</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/routers/sessions/chat.py

<li>Integrated <code>tool_calls_evaluator</code> into the <code>chat</code> function.<br> <li> Replaced direct <code>litellm.acompletion</code> calls with evaluated tool calls.<br> <li> Enhanced modularity by delegating tool call handling.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1179/files#diff-37385dc5ce1960a145f02585085ed045b6e53758271e5107a16a09b6e2653b7c">+7/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>tools.py</strong><dd><code>Introduce tool call evaluation utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/agents_api/routers/utils/tools.py

<li>Added <code>tool_calls_evaluator</code> decorator for tool call evaluation.<br> <li> Implemented <code>call_tool</code> function to handle system tool calls.<br> <li> Defined <code>_system_tool_handlers</code> mapping for tool-specific logic.<br> <li> Added utility functions for search request creation and tool handling.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1179/files#diff-d26f4f93dfddca41949d5d2b72a761fab0ae3496b803606e0b981dffd316626a">+268/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_tool_calls_evaluator.py</strong><dd><code>Add unit tests for tool evaluation utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

agents-api/tests/test_tool_calls_evaluator.py

<li>Added unit tests for <code>call_tool</code> covering various tool operations.<br> <li> Tested <code>tool_calls_evaluator</code> for correct tool call handling.<br> <li> Included edge cases like unknown tools and non-matching tool types.<br> <li> Verified iterative tool call evaluation until completion.

</details>

  </td>
  <td><a href="https://github.com/julep-ai/julep/pull/1179/files#diff-d008fdc91819df58a8de0ceceda7fb5ea583fcc5832096ba6a9ecfd58e42a0ed">+524/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

----

> [!IMPORTANT]
> Refactor chat endpoint to use `tool_calls_evaluator` for tool call handling, enhancing modularity and adding extensive tests.
> 
>   - **Behavior**:
>     - Refactor `chat()` in `chat.py` to use `tool_calls_evaluator` for handling tool calls, replacing direct `litellm.acompletion` calls.
>     - Introduce `call_tool` in `tools.py` for executing system tool calls.
>   - **Utilities**:
>     - Add `tool_calls_evaluator` decorator in `tools.py` for tool call evaluation.
>     - Define `_system_tool_handlers` mapping in `tools.py` for tool-specific logic.
>   - **Tests**:
>     - Add `test_tool_calls_evaluator.py` with unit tests for `tool_calls_evaluator` and `call_tool`.
>     - Cover various tool operations and edge cases like unknown tools.
>   - **Misc**:
>     - Add utility functions for search request creation in `tools.py`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=julep-ai%2Fjulep&utm_source=github&utm_medium=referral)<sup> for ebd3769ebebdd0d2aa05af0eba020399b93c0e25. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: agents-api/tests/test_tool_calls_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,1807 @@
...
+user_doc_list = make_acompletion_multiple_outputs([
+    {
+        "role": "assistant",
+        "content": "Listing user docs",
+        "tool_calls": [
+            {
+                "id": "call_user_doc_list",
+                "type": "system",
+                "function": {
+                    "name": "user.doc.list",
```
Comment (creatorrr) [inline agents-api/tests/test_tool_calls_evaluator.py:267]: I think this (and related) should be `user.docs.list` etc to keep things consistent with the api and sdk

File: agents-api/tests/test_tool_calls_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,1807 @@
...
+        ],
+    },
+    {
+        "role": "assistant",
+        "content": "Here are the agent docs",
+    },
+])
+
+agent_doc_create = make_acompletion_multiple_outputs(
+    lambda agent, doc, user, task, user_doc, session: [
```
Comment (creatorrr) [inline agents-api/tests/test_tool_calls_evaluator.py:44]: could be simplified as `lambda agent, **_: ...`

File: agents-api/tests/test_tool_calls_evaluator.py [hunk 3]
```diff
@@ -0,0 +1,1807 @@
...
+            ],
+        },
+        {
+            "role": "assistant",
+            "content": "User doc deleted successfully",
+        },
+    ]
+)
+
+user_doc_search = make_acompletion_multiple_outputs([
```
Comment (creatorrr) [inline agents-api/tests/test_tool_calls_evaluator.py:338]: we could probably further DRY this whole stuff with:

```python
def example_messages(op_name: str, target: str, **arguments):
    return [
				{
				        "role": "assistant",
				        "content": f"{op_name}-ing {target}",
				        "tool_calls": [
				            {
				                "id": f"call_{target.replace(' ', '_')}_{op_name}",
				                "type": "system",
				                "function": {
				                    "name": f"{target.replace(' ', '.')}.{op_name}",
				                    "arguments": {
				                        "x_developer_id": str(UUID(int=0)),
				                        **arguments,
				                    },
				                },
				            }
				        ],
				    },
				    {
				        "role": "assistant",
				        "content": "Here are the results",
				    },
				]

# then:
user_doc_search = make_acompletion_multiple_outputs(example_messages('search', 'user docs', user_id=str(uuid7()), search_params={"text": "test query"})
```

File: agents-api/tests/test_tool_calls_evaluator.py [hunk 4]
```diff
@@ -0,0 +1,1807 @@
...
+
+@test("chat: evaluate agent.doc.list tool call")
+async def _(
+    make_request=make_request,
+    developer_id=test_developer_id,
+    agent=test_agent,
+    mocks=agent_doc_list,
+    dsn=pg_dsn,
+):
+    pool = await create_db_pool(dsn=dsn)
```
Comment (creatorrr) [inline agents-api/tests/test_tool_calls_evaluator.py:752]: shouldn't the correct test be like:

```python
    test_agent = client.agents.create(..., tools=[dict(type='system', ...)])  # add agent.docs.list system tool
    # ...
    mocks = ... #  from test_agent.tools?
    
    response = client.sessions.chat(agent=test_agent, ...)  # this should call the mocks
    # other tests, asserts etc
```

File: agents-api/tests/test_tool_calls_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,1807 @@
...
+                "max_query_length": 1001,
+                "confidence": 0.6,
+                "limit": 5,
+                "mmr_strength": 0.5,
+            },
+        ),
+        connection_pool=pool,
+    )
+
+    (_embed, acompletion) = mocks
```
Comment (creatorrr) [inline agents-api/tests/test_tool_calls_evaluator.py:884]: maybe something like:

```python
   test_agent = client.agents.create(...,
       tools=[dict(name='search_docs', type='system', resource='agent.docs', op='search')]
    )

    with make_acompletion_from_tools(agent.tools) as (_embed, acompletion):
        # since agent.tools has the search_docs function, this should have called agent.docs.search
        ...
```

File: agents-api/agents_api/routers/utils/tools.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+
+                            tc = tool_calls[tool_call.index]
+                            if tool_call.id:
+                                tc["id"] = tool_call.id
+                            if tool_call.function.name:
+                                tc["function"]["name"] += tool_call.function.name
+                            if tool_call.function.arguments:
+                                tc["function"]["arguments"] += tool_call.function.arguments
+
+            for tool in tool_calls:
```
Comment (creatorrr) [inline agents-api/agents_api/routers/utils/tools.py:345]: this should be parallelized using `await asyncio.gather(*...)`

**Comment by creatorrr** [comment]
@whiterabbit1983 to clarify further (this bit is from claude's pr review):

> 
>   In the current PR implementation, the chat endpoint sets up the ToolCallsEvaluator with:
> 
>   evaluator = ToolCallsEvaluator(
>       completion_func=litellm.acompletion,
>   )
>   model_response = await evaluator.completion(developer.id, **payload)
> 
>   The key concern is that the tool call evaluation seems to be working at a system level rather than being tied to the
>   specific tools defined for each agent. In your expected behavior:
> 
>   1. agent1 with no tools should never have access to system tools
>   2. agent2 with specific tools should only have access to those tools
> 
>   I don't see where this filtering happens in the current implementation. The tools for the model are set in the
>   render_chat_input function which creates formatted_tools, but there doesn't appear to be proper connection between the
>   tools provided to the model and the tools that the evaluator will actually execute.
> 
>   Additionally, there seems to be a type mismatch between what the agent configuration shows (type='system') and what the LLM
>    expects (type='function'), which might explain why some tests are failing.
> 
>   This implementation needs to be updated to ensure:
>   1. Only tools configured for a specific agent are available in the chat
>   2. The tool types match between configuration and LLM expectations

File: agents-api/tests/test_tool_calls_evaluator.py (hunk 3 above)
Comment (whiterabbit1983) [reply agents-api/tests/test_tool_calls_evaluator.py:338]: it's allowed to be repetitive for tests

File: agents-api/tests/test_tool_calls_evaluator.py (hunk 4 above)
Comment (whiterabbit1983) [reply agents-api/tests/test_tool_calls_evaluator.py:752]: this involves the client, which might change interface in the future

**Comment by creatorrr** [comment]
Julep's main branch is now Julep 3, a ground-up rewrite (see the pinned announcement #1610). v1 is preserved on the v1 branch and its docs at v1.docs.julep.ai. Closing v1-era items; please re-file against v3 if still relevant.
